lib/puppet/indirector/file_bucket_file/file.rb

```diff
@@ -189,19 +210,45 @@ def path_for(bucket_path, digest, subfile = nil)
       Puppet::FileSystem.pathname(subfile ? ::File.join(basedir, subfile) : basedir)
     end
 
-    # @param contents_file [Object] Opaque file path
-    # @param bucket_file [IO]
-    def verify_identical_file!(contents_file, bucket_file)
-      if bucket_file.size == Puppet::FileSystem.size(contents_file)
-        if bucket_file.stream() {|s| Puppet::FileSystem.compare_stream(contents_file, s) }
-          Puppet.info "FileBucket got a duplicate file #{bucket_file.checksum}"
-          return
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param bucket_file [Puppet::FileBucket::File] IO object representing
+    #   content to back up
+    # @return [Boolean] whether the data in contents_file is of the same size
+    #   and content as that in the bucket_file
+    # @api private
+    def verify_identical_file(contents_file, bucket_file)
+      (bucket_file.size == Puppet::FileSystem.size(contents_file)) &&
+        (bucket_file.stream() {|s| Puppet::FileSystem.compare_stream(contents_file, s) })
+    end
+
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param expected_checksum_data [String] expected value of checksum of type
+    #   checksum_type
+    # @param checksum_type [String] type of check sum of checksum_data, ie "md5"
+    # @return [Boolean] whether the checksum of the contents_file matches the
+    #   supplied checksum
+    # @api private
+    def contents_file_matches_checksum?(contents_file, expected_checksum_data, checksum_type)
+      contents_file_checksum_data = Puppet::Util::Checksums.method(:"#{checksum_type}_file").call(contents_file.to_path)
+      contents_file_checksum_data == expected_checksum_data
+    end
+
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param bucket_file [Puppet::FileBucket::File] IO object representing
+    #   content to back up
+    # @return [void]
+    # @api private
+    def copy_bucket_file_to_contents_file(contents_file, bucket_file)
+      Puppet::Util.replace_file(contents_file, 0440) do |of|
```

Comment:
After investigation, it turns out that for the tempfile that it creates, `Puppet::Util::replace_file` will either

  a) use the mode of an existing target file as the mode of the tempfile
or
  b) use the supplied mode argument if the target file does not exist

But it chmods the tempfile to this mode _before_ yielding to the block defined by the caller.

This is unintuitive for a couple reasons. First, it makes more sense to me to have the supplied mode override the mode of an existing file (if I supply an explicit argument, it should be honored). Second, chmoding the tempfile before yielding means a read-only tempfile?

For the purposes of this PR, this then plays out thusly:

```ruby
Puppet::Util.replace_file('/tmp/foo', 0440) { |of| FileUtils.copy_stream(StringIO.new("\u1234"), of) }
```
If a `contents` file does not exist (which we expect to happen a substantial percentage of the time), `Puppet::Util::replace_file` will create a tempfile with a mode of 0440 (read only), and then supply this handle to the block. Since we can't actually write to this read-only tempfile, the copy fails.

If a `contents` file does exist (which will happen any time there is a duplicate entry, a substantial percentage also), it _should_ have a mode of 0440. So then `Puppet::Util::replace_file`, when seeking to overwrite the `content`s file with a new one, will create a tempfile with a mode of 0440, and we will fail to overwrite it.

I suppose we could wrap the call to `Puppet::Util::replace_file` with a preceding `chmod 0640` on the `contents` file and succeeding `chmod 0440` afterward, but this seems to be a dubious direction...what happens if we chmod 640 and then die during the write? It seems to me a better change would be to handle this use-case in `Puppet::Util::replace_file`.

I'd prefer not to hold up this PR for that so I'm inclined to remove the commit that adds replace_file from the PR.